### ISSUE TITLE
Finalizer problem

### ISSUE THREAD

**Issue by Ilir-Liburn** [HaxeFoundation/hxcpp#1001]
Hello,

I have problem using finalizer and need guidance. Here is intro:

Before: 0
After: 16000016
Copied: 32000032
FINALIZER!!!
Destroyed: 16000528
Collected: 512

After: allocation of the first array
Copied: allocation of the second (copied!) array
FINALIZER!!!: triggered while destroying second copied array
Destroyed: allocation after second array is destroyed
Collected: allocation after first array is destroyed

The problem is: finalizer triggers while destroying copied array, instead of the first array where finalizer is given. I use Variant as element type (need this because of some extern data not allocated by HxCPP, where object to be finalized is stored in valObject field.

I tried this using alloc_empty_object->val_gc, switched to CreateDynamicPointer later because it returns hx::PointerData on extern data which is just what I need.

My question is:  How to get finalizer triggered on last pointer reference to the object? Am I doing something wrong?

And one more question: How can I get custom element type defined? Variant is OK, but I can also make what I need using union of Double/Int64/Pointer (no type, string len needed)

**Comment by Ilir-Liburn** [comment]
Finalizer is called as soon I call __hxcpp_collect (right after creation of the first array). Therefore, finalizer is not linked to the object stored in array field, nothing I can do here.

**Comment by Ilir-Liburn** [comment]
Stand corrected, I managed to do what I need by attaching finalizer to array(s) instead of items. And it works perfectly: finalizer is called by the last array destroyed (when there are no more references), no matter if last array destroyed was copied or not.

One thing: following code doesn't work:

```C
	::Array_obj<cpp::Variant> *obj = (::Array_obj<cpp::Variant> *)arr;
	obj->__SetSizeExact(0);
	delete obj;
```

First I noticed delete obj is not destroying array, so I had to use SetSizeExact. Then I realized finalizer is not called (because object was not destroyed), so I ended up using

```C
	*(((int *)(arr))-1) = 0;
```
After clearing pointer header, array is properly collected, finalizer is called and memory is wiped out. Wonder if there is a better way....
